Fix rectified receipt persistence and bump koinos-chain to v1.5.1 (#858)
koinos/koinos-chain: RECEIPT_PERSISTENCE_FIX_AND_RECOVERY.md

# Receipt Persistence Fix And Recovery Notes

## TL;DR

This branch contains the minimal code fix for the rectification receipt persistence bug in `koinos-chain`.

The bug was that rectified block receipts were computed, but the original unrectified receipt was the one persisted to `block_store` and broadcast through `koinos.block.accept`.

That means:

- the fix in this branch prevents future bad receipts from being written,
- but it does not retroactively repair block stores that already contain corrupted historical receipts.

For already-affected nodes, `chain.verify-blocks=true` is still required during reindex so that the node re-executes blocks instead of trusting persisted deltas.

> This branch also bumps the chain version to `1.4.2`.

Reviewer: ```suggestion
This branch also bumps the chain version to `1.5.1`.
```

## Version

This branch changes:

> - `koinos-chain` `1.4.1 -> 1.4.2`

Reviewer: ```suggestion
- `koinos-chain` `1.4.1 -> 1.5.1`
```

### Case 2: Existing affected nodes

If a node already synced through the affected window and persisted bad receipts, the recommended recovery path is:

1. Upgrade to a binary containing this fix.
2. Restart with `chain.verify-blocks=true`.
3. Let the node reindex by re-executing blocks.
4. Once the node has rebuilt state from execution and is healthy again, decide whether to keep or remove `verify-blocks` depending on operational preference.

The key point is that step 2 is needed because the node's historical receipt data is already wrong.

## What This Branch Does Not Attempt To Solve

This branch is intentionally scoped.

It does not:

- mutate old receipts already stored in `block_store`,
- introduce a fallback path inside `apply_block_delta()`,
- automatically detect old corrupted receipts and self-heal,
- change network protocol behavior.

It only fixes the persistence bug at the point where the wrong receipt was being written.

## Branch Intent

This branch is meant to stay clean and isolated relative to `master`:

- rectification receipt persistence fix,

> - patch version bump to `1.4.2`,

Reviewer: ```suggestion
- patch version bump to `1.5.1`,
```
